README
WARSYNC v0.9.9
    Warsync is the program that initiates replication between the Warsync
    server and it's clients. Warsync is designed to be called from the
    command line of the Warsync server to initiate replication on demand. It
    can also be called from cron on the server to automate replication.

    Warsync not only replicates files from the server to the client, but can
    also synchronize what software packages are installed on a Debian
    GNU/Linux system.

    Warsync determines what files to replicate by processing the
    /etc/warsync/filepacks.d directory on the server. This directory
    contains *filepacks* which are simple text files that contain lists of
    the files and directories that should be replicated.

Download
    Warsync can be downloaded from the Warsync Alioth project page at
    <http://alioth.debian.org/projects/warsync/>

    It is available as a Debian package (.deb) and as a source tarball
    (.tar.gz)

    It is also available as an apt-get source using the following entry:

     deb http://givetheworld.net/debian sarge warsync
     deb-src http://givetheworld.net/debian sarge warsync

Install
    On a Debian GNU/Linux system, the best way to install Warsync is to add
    the apt sources lines above to your /etc/apt/sources.list file. You can
    then be able to install Warsync with the following commands:

     apt-get update
     apt-get install warsync

    Alternatively you can download the Debian package directly and install
    it with dpkg.

     dpkg -i warsync_0.9.9_all.deb

    If you receive an error about missing dependencies, a quick call to
    apt-get will fix that up.

     apt-get -f install

    On other Linux distributions, or if you are a glutton for punishment,
    Warsync can also be installed manually from the source tarball.
    Although, this method has not been tested. A from source installation
    would go something like this:

     tar zxvf warsync_0.9.9.tar.gz
     cd warsync-0.9.9
     perl Makefile.PL
     make
     make install

Warsync Configuration
    First configure the Warsync server by executing the following command on
    the system that will serve as the Warsync server.

     warsync-config --server

    Next configure Warsync clients by executing the following command on
    each Warsync client.

     warsync-config --client

    Back on the Warsync server, execute the following command to fully
    register each client system with the server.

     warsync-add-client clientname

    After this you will need to create *filepacks* in the
    /etc/warsync/filepacks.d directory on the Warsync server. Each filepack
    should contain a list of directories or files to be replicated when the
    filepack name is specified on the warsync command line.

Filepacks
    Filepack should list each file or directory pattern to be replicated one
    per line. It is best to have several filepacks where each filepack
    replicates a group of related files and directories. For instance you
    may have one filepack called homes that will replicate user home
    directories. You could have another filepack called website that will
    replicate the files for your website. Another filepack called config
    could be used to copy configuration or "/etc" files that you want to
    keep replicated between server and client.

    To match a file, specify the full absolute path to the file. For
    example, "/etc/passwd". Wildcards "?" and "*" may also be specified. "?"
    matches any single character, while "*" will match zero or more
    characters. For example, "/etc/pa*" could match "/etc/pam.conf" and
    "/etc/passwd".

    To replicate a directory and all it's subdirectories, specify two
    asterisk, "**" after the directory name. For example "/home/**" will
    replicate everything in the "/home" directory. If only one asterisk is
    specified, then only the files directly contained in the directory will
    be replicated. The subdirectories will not be replicated.

    To exclude files from being replicated that would match the wild card in
    another entry, specify them on line before the include line. It should
    start with "- " (a minus followed by a space). For example,
    "- /etc/network/interfaces" followed by "/etc/network/**" would
    replicate all files and subdirectories contained in the "/etc/network"
    directory, except the "interfaces" file.

    Filepacks may also contain comments and are signified when the first
    non-whitespace character on a line is a "#".

    An example homes filepack could look like:

     # replicate home directories.
     /home/**
     # replicate root's home, but not it's .ssh directory
     # otherwise warsync will probably break.
     - /root/.ssh
     /root/**

    Example website filepack could be as simple as:

     /var/www/**

    A config filepack for replicating config files could look like:

     # copy user info
     /etc/passwd
     /etc/shadow
     /etc/group
     /etc/gshadow
     /etc/skel/**
     # ssh config
     - /etc/ssh/ssh_host_dsa_key*
     - /etc/ssh/ssh_host_rsa_key*
     /etc/ssh/**

    Some directories that you probably never want to completely replicate
    any top level directories, especially /boot, /cdrom, /dev, /floppy,
    /initrd, /lost+found, /mnt, /proc, /tmp, and /var. You also probably
    shouldn't replicate /bin, /lib, /sbin, or /usr as these directories
    should be managed by your operating systems package management system.
    Although replicating /usr/local shouldn't hurt. While you may want to
    replicate some of the files in /etc, you definitely do NOT want to
    replicate them all. Also replicating the /etc/warsync directory will
    break configuration on the clients. The following directories are
    sensitive to warsync's operation and should not be replicated unless you
    absolutely know what you are doing:

     /etc/warsync
     /etc/ssh
     /root/.ssh
     /var/log/warsync

    Warsync makes no effort to keep you from replicating things you that you
    shouldn't. A good rule of thumb is to be very explicit about what to
    replicate and use fewer wildcards rather than more. And if you don't
    know what a file does, don't replicate it.

Running Warsync
    To replicate a single filepack, specify it on the warsync command line
    like so:

     warsync foobar

    The above command will replicate the files and directories specified in
    /etc/warsync/filepacks.d/foobar file pack. Multiple filepacks can be
    specified at once to replicate several filepacks one after the other:

     warsync homes website

    This would replicate the homes filepack, and then the website file pack.

    To replicate all filepacks, specify "--all-filepacks" on the command
    line:

     warsync --all-filepacks

    You can also use "-a" for short to replicate all filepacks:

     warsync -a

    If you want to test a filepack to see what will be replicated before
    actually making any real live changes, you can specify "--dry-run" on
    the command line:

     warsync --dry-run config

    This will simulate replication of the config filepack. You can also use
    "-n" for short instead of "--dry-run".

     warsync -n foobar

    If you want to see more verbose output about what is happening then
    specify "--verbose" on the command line:

     warsync --verbose homes

    You can also specify "-v" as a shortcut. To see even more output, you
    can specify multiple "v"'s like so:

     warsync -vv config

    And of course this can be combined with "-n" to simulate a replication
    with extra output:

     warsync -vn --all-filepacks

    To completely limit the output produced by warsync so that it will only
    display output if an error occurs, add "--quiet" to the command line:

     warsync --quiet homes

    And of course "--quiet" can be shortened to just "-q" as in:

     warsync -q foobar

    By default warsync will replicate specified filepacks to all clients. If
    you only want to replicate to a specific warsync client, you can specify
    the "--client" option on the warsync command line;

     warsync --client=web2 website

    The above command would replicate the website filepack to only the
    client named web2. More than one client can be specified at a time by
    using multiple "--client" options:

     warsync --client=web2 --client=web9 website

    Warsync can also synchronize the software packages installed on Debian
    GNU/Linux systems. The Debian package synchronization will install,
    remove, upgrade, and/or *downgrade* the Debian packages on a Warsync
    client in order to match the packages installed on the server.

    To synchronize the Debian packages installed on the client with those on
    the server, specify the "--debian" option on the warsync command line.

     warsync --debian

    If "--debian" is specified along with filepacks, the filepacks will be
    replicated first before doing the Debian package synchronization.

    Warsync can also be used to take a snapshot of installed Debian packages
    and then later used to "restore" back to this state. This comes in handy
    if you want to install a package that requires several libraries that
    you don't want to leave installed if you later decide to remove the
    package. Prior to installing the new package, take a snapshot of
    installed Debian packages by executing:

     warsync --debian-snapshot=debsnap

    This will write the snapshot to the debsnap file in the current
    directory. Later when you want to restore back to package state, run the
    command:

     warsync --debian-from-snapshot=debsnap

    The debsnap file can also be copied to another system with Warsync
    installed to install the same packages on another system.

Automating Warsync Via Cron
    Warsync can be comfortably run from a crontab for filepack replication.
    In fact warsync does special handling when it is not run from a tty,
    such as when run from cron, and will not display any output unless
    changes are made, or "-v" has been specified on the command line. In
    this fashion, when warsync runs from cron, you will not receive a cron
    email unless changes were actually made on the server. This helps to cut
    down on the noise level if you want to run warsync periodically through
    out the day, but only want to receive a notification when a change was
    made. If you always want to receive output even when no changes were
    made, specify "-v" on the warsync command line in the crontab. If you
    want to receive no output at all, specify "-q" on the command line. In
    this case you will only receive an email if there was an error.

    It should be noted that running Warsync from cron to do Debian package
    synchronization has not been extensively tested and should be considered
    experimental. This is mainly because when a new package is installed or
    upgraded, you may be prompted to answer configuration questions, which
    can yield unpredictable results when running from cron.

    Because warsync can replicate individual filepacks at a time, it is
    beneficial to put multiple cron entries that replicate different
    filepacks throughout the day. For instance, you might want to
    automatically replicate changes to the website every thirty minutes, but
    only replicate user home directories every two hours. And you may want
    to do a full replication every night at 11 o'clock. In that case you
    might have a crontab that looks like this:

     15,45 * * * * /usr/sbin/warsync website
     30 */2 * * * /usr/sbin/warsync homes
     0 23 * * * /usr/sbin/warsync -a

    Of course it might get pretty annoying to get emailed about every single
    change made to users' home directories. To quiet the noise, you'll want
    to change the second crontab line as shown below.

     30 */2 * * * /usr/sbin/warsync -q homes

Advanced Configuration
    For those of you that really want to get the most out of Warsync, there
    are some advanced configuration options you can specify in the
    server.conf file and within filepacks.

  server.conf Options
    While filepacks tell Warsync *what* to replicate, the server.conf file
    tells Warsync *where* and *how* to perform its replication. The
    server.conf file stores the list of Warsync clients and different
    options that apply to each individual client. The server.conf file uses
    the INI config style that is prevalent on Microsoft Windows and also
    used in MySQL. Each client's config is started off with a config
    grouping line in the form of "[client-*clientname*]". Then each
    "name=value" line afterwards is applied to the client until the next
    config group line is reached. Comments may be included in the
    server.conf just like in filepacks. Comments are preceeded by "#"
    character.

    A typical server.conf might look like this:

     [client-web2]
     fqdn = web2.mydomain.com
     port = 22
     compress = 0
     skip = 0
 
     [client-web3]
     fqdn = web3.mydomain.com
     port = 22
     compress = 0
     skip = 0

    The above config file lists to Warsync clients, web2 and web3. It also
    specifies their fully-qualified domain names, what port to connect over
    for ssh, that compression should not be used, and neither one is skipped
    by default. These four options are configured whenever you run

debian/changelog
warsync (0.9.9~bayru2) unstable; urgency=low

  * Fix install path

 -- jlawler <jlawler@morgoth.crazygeek.org>  Wed, 04 Jan 2012 01:26:58 -0600

warsync (0.9.9~bayru1) unstable; urgency=low

  * Non-maintainer upload.
  * Some debian packages have a ~ in the version string
  * Perl's File::Temp might return _ in the filename

 -- jlawler <jlawler@morgoth.crazygeek.org>  Wed, 04 Jan 2012 01:03:27 -0600

warsync (0.9.8+rc0.9.9+4) unstable; urgency=low

  * This is version 0.9.9 Release-Candidate 4 and new features may not work as
    expected.
  * When replicating all servers, they are now replicated in alphabetic order
    instead of a seemingly random order.
  * Introduced -t parameter to prefix output with timestamps.
  * Other minor improvements and bug fixes.

 -- Paul Baker <paulb@feedburner.com>  Wed, 20 Sep 2006 17:45:20 -0500

warsync (0.9.8+rc0.9.9+3) unstable; urgency=low

  * This is version 0.9.9 Release-Candidate 3 and new features may not work
    as expected.
  * Dynamic client-side exclude support implemented with .warsyncignore files.
  * Client-side pre, mod, and post command support implemented for filepacks.
    Pre commands run on the client prior to syncing. Mod commands run after
    syncing only if modifications were made. Post commands run after syncing
    even if no modifications were made.
  * Warsync event tracking implemented so multiple server processes will not
    try to perform the same operation on a client.
  * If the installed version of rsync supports it (v2.6+), rsync process will be
    restarted when files disappear from the server during rsync instead of
    causing an error.
  * If running rsync v2.6.4 or higher, files will be deleted from the client
    incrementally (using --delete-during rsync option) instead of at the
    beginning of the rsync process. This will speed up the rsync process. See
    documentation for reverting back to old behavior.
  * Server will no longer replicate all clients if a non-existant client is
    specified on the command line.
  * When adding new clients to a server, user is asked if they want client to
    be skipped by default when replicating all clients.
  * Specifying RSYNC_OPT options in a filepack actually works now.
  * New filepack option DELETE when set to a false value, the filepack will not
    delete files from the warsync client that do not exist on the server.
  * New filepack option SKIP will cause a filepack to not be included when
    running warsync --all-filepacks.
  * Many minor improvments and bug fixes.

 -- Paul Baker <paulb@feedburner.com>  Thu, 23 Feb 2006 13:53:29 -0800

warsync (0.9.8) unstable; urgency=low

  * Filepack replication introduced.
  * Debian package snapshot improvements.
  * Output improvments. -q and -v actually mean something along with special
    handling when run from non-tty (such as cron).
  * Real documentation! See warsync-howto manpage.

 -- Paul Baker (w2gi) <pbaker@where2getit.com>  Mon, 10 May 2004 15:57:25 -0500

warsync (0.9.7) unstable; urgency=low

  * New Debian Package snapshot feature.
  * Fairly substantial refactoring of code resulting in much less redundant
    code.
  * Significant speed improvment in calculating differences in packages
    between client and server.
  * More flexible "ini" style config file for server.

 -- Paul Baker (w2gi) <pbaker@where2getit.com>  Sat, 27 Mar 2004 14:50:47 -0600

warsync (0.9.6.2) unstable; urgency=low

  * New method of purging packages from client that works when a package has
    already been removed, but not purged.
  * Improved debian/rules.
  * Fixes to debian/copyright file.
  * Numerous manpage improvements.

 -- Paul Baker (w2gi) <pbaker@where2getit.com>  Fri, 13 Feb 2004 18:21:32 -0600

warsync (0.9.6.1) unstable; urgency=low

  * Forgot to install warsync-add-client and warsync-client-push.

 -- Paul Baker (w2gi) <pbaker@where2getit.com>  Wed,  4 Feb 2004 09:07:02 -0600

warsync (0.9.6) unstable; urgency=low

  * First alpha release of Warsync.

 -- Paul Baker (w2gi) <pbaker@where2getit.com>  Wed,  4 Feb 2004 08:55:09 -0600

pfars (0.9.5) unstable; urgency=low

  * new COMPRESS config option to enable compression between server
    and client. disabled by default.
  * new CIPHER config option to specify cipher used to encrypt
    communication between server and client. blowfish used by
    default.
  * default location for client identity file changed to
    /etc/pfars/private/id_dsa.
  * more verbose output by default.

 -- Paul Baker (w2gi) <pbaker@where2getit.com>  Sun, 1 Jun 2003 06:41:24 -0500

pfars (0.9.4) unstable; urgency=low

  * The "Operation Spirit (The Tyranny of Tradition)" Release.
  * fixes mistake in documentation reported in sf.net bug#596736.
  * implements new .pfarsignore file support. disabled by default.
  * finally fixed restart when pfars is updated during replication.

 -- Paul Baker (w2gi) <pbaker@where2getit.com>  Tue, 24 Sep 2002 11:06:35 -0500

pfars (0.9.3) unstable; urgency=low

  * The "Sympathy" Release.
  * actually make the init-files.local and files.local files useful.
  * new packages.local file for overriding installed debian packages.
  * improved default init-files config file to fix bugs and include
    /etc/apt/sources.list.
  * corrected debian/rules binary-* targets.

 -- Paul Baker (w2gi) <pbaker@where2getit.com>  Sun, 11 Aug 2002 23:01:18 -0500

pfars (0.9.2.2) unstable; urgency=low

  * The "Don't Let Me Down" Release.
  * how about we install that dandy little cronjob file included
    in the package.

 -- Paul Baker (w2gi) <pbaker@where2getit.com>  Thu, 18 Jul 2002 06:23:59 -0500

pfars (0.9.2.1) unstable; urgency=low

  * The "Inspection Wise 1999" Release.
  * fixed log file output.
  * hopeful fix for restart failure on pfars upgrade.

 -- Paul Baker (w2gi) <pbaker@where2getit.com>  Thu, 18 Jul 2002 05:54:59 -0500

pfars (0.9.2) unstable; urgency=low

  * The "Campfire Kansas" Release.

lib/Warsync/Host.pm
package Warsync::Host;

##############################################################################
## Warsync::Host - Warsync                                                  ##
##                                                                          ##
## Copyright (C) 2001-2006 Paul J. Baker                                    ##
##                                                                          ##
## This program is free software; you can redistribute it and/or modify     ##
## it under the terms of the GNU General Public License version 2 as        ##
## published by the Free Software Foundation.                               ##
##                                                                          ##
## This program is distributed in the hope that it will be useful,          ##
## but WITHOUT ANY WARRANTY; without even the implied warranty of           ##
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the            ##
## GNU General Public License for more details.                             ##
##                                                                          ##
## You should have received a copy of the GNU General Public License        ##
## along with this program; if not, write to the Free Software              ##
## Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.                ##
##                                                                          ##
## You may contact the author via electronic mail at warsync@paulbaker.net  ##
##############################################################################

## $Id: Host.pm 174 2006-09-20 22:39:21Z pbaker-guest $

=head1 NAME

Warsync - Wrapper Around Rsync

=head1 SYNOPSIS

  package Warsync::Foo;

  use Warsync::Host;

  our @ISA = qw( Warsync::Host );

=head1 DESCRIPTION

Common methods for Warsync hosts.

=head1 AUTHOR

Paul Baker, pbaker@where2getit.com

=head1 SEE ALSO

warsync(8).

=cut

#############################################################################

require 5.6.0;
use strict;
use warnings;

use Carp;
use Data::Dumper;
use Getopt::Long ();
use Pod::Usage;
use Sys::Hostname;
use POSIX 'strftime';

use Warsync::Util;

require Exporter;

our @ISA = qw( Exporter );

sub dryrun
{
    my $self = shift->new;

    return $self->option('n') || 0;
}

sub option
{
    my $self = shift->new;
    my $key  = shift;

    $self->options->{$key};
}

sub options
{
    my $self = shift->new;

    if (@_) {
        $self->{options} = shift;
        return $self;
    }

    return $self->{options};
}

sub display
{
    my $self  = shift->new;

    ## make sure they gave us at least two arguments.
    croak 'usage: $OBJECT->display(LEVEL, LIST)' unless @_ > 1;

    ## check for timestamp.
    my $time = \time;
    if (ref($_[0]) eq 'SCALAR') {
        $time = shift;
    }

    if (my $pat = $self->timestamp) {
        print STDERR strftime($pat, localtime $$time),' ', @_ if $self->verbose >= shift;
    }
    else {
        print STDERR @_ if $self->verbose >= shift;
    }
}

sub packages
{
    my $self = shift->new;

    $self->{_packages} ||= $self->debian_packages;

    return deep_copy $self->{_packages};
}

sub parse_cli_params
{
    my $self = shift->new;

    my %option = ();

    ## set options for host object.
    $self->options(\%option);

    ## set option defaults.
    Getopt::Long::Configure("bundling");

    Getopt::Long::GetOptions(
        \%option,
        'help|H|?',
        'man',
        'version|V',
        'v|verbose+',
        'q|quiet',
        'n|dry-run',
        't|timestamp',
        @_
    ) or pod2usage(2);

    pod2usage(1) if $option{help};
    pod2usage(-exitstatus => 0, -verbose => 2) if $option{man};

    # display version number if --version set.
    if ($option{version}) {
        $self->display(0, version);
        exit;
    }

    ## always be at least little verbose.
    ++$option{v};
    ## but be quiet if asked.
    $option{v} = 0 if $option{'q'};
    
    ## blow out skip-*-cmds options.
    if ($option{'skip-cmds'}) {
        $option{'skip-mod-cmds'}  = 1;
        $option{'skip-pre-cmds'}  = 1;
        $option{'skip-post-cmds'} = 1;
    }

    return $self;
}

sub name
{
    my $self = shift->new;

    return Sys::Hostname::hostname;
}

sub replication
{
    my $self = shift->new;
    if (tty) {
        $self->display(0, "This is a Warsync client. Warsync replication must be\n");
        $self->display(0, "initiated from the Warsync server only.\n");
        $self->exit_code(1);
    }
    return;
}

sub run
{
    my $self = shift->new;

    $self->parse_cli_params(
        'a|all-filepacks',
        'include-skipped-filepacks|i',
        'client=s@',
        'debian',
        'skip-mod-cmds|m',
        'skip-pre-cmds',
        'skip-post-cmds',
        'skip-cmds',
        'debian-closest-version',
        'debian-ignore=s@',
        'debian-snapshot=s',
        'debian-from-snapshot=s',
        'check-clients',
        'accept-unknown-hosts',
        'client-offset=s',
    );

    ## if --debian-snapshot argument given, do that.
    if ($self->option('debian-snapshot')) {
        $self->debian_snapshot;
        exit($self->exit_code);
    }

    ## if --debian-from-snapshot argument given, do that.
    if ($self->option('debian-from-snapshot')) {
        $self->debian_from_snapshot;
        exit($self->exit_code);
    }

    #print STDERR Data::Dumper::Dumper($self->options);

    $self->replication;
}

sub debian_packages
{
    my $self = shift->new;

    my $v = $self->verbose;

    ## get current package information.
    my $packages;

    {
        my $prev_col = $ENV{COLUMNS};
        ## now set it to 200
        $ENV{COLUMNS} = 200;

        ## read dpkg -l
        open(my $dpkg_h, 'dpkg -l |')
            or die "could not exec dpkg -l: $!";

        $packages = parse_package_list($dpkg_h);

        close($dpkg_h);

        if ($prev_col) {
            $ENV{COLUMNS} = $prev_col;
        }
        else {
            delete $ENV{COLUMNS};
        }
    }

    my $total   = keys(%$packages);
    my $done    = 0;
    my $percent = 0;

    ## iterate through packages and get real version number.
    {
        open(my $fh, "</var/lib/dpkg/status")
            or $self->fault("Could not read /var/lib/dpkg/status: $!");

        my $package = '';
        while (<$fh>) {
            chomp;
            next unless length;

            ## current package block.
            if (my ($pkg) = $_ =~ /^Package:\s+(\S+)/) {
                $package = $pkg;
                next;
            }

            next unless exists $packages->{$package};

            if (my ($ver) = $_ =~ /^Version:\s+(\S+)/) {
                $packages->{$package}[2] = $ver;

                ## only display progress if verbose is on
                ## and we are on a tty.
                if ($v and tty) {
                    my $new = int( (++$done / $total) * 100);
                    if ($new > $percent) {
                        my $str = "$new%";
                        $self->display(1, "$str", "\b" x length($str));
                        $percent = $new;
                    }
                }
            }
        }

        close($fh);
    }

    if ($v and tty) {
        $self->display(1, ' ' x 5, "\b" x 5);
    }

    return $packages;
}

sub debian_snapshot ($)
{
    my $self = shift->new;

    ## open file to write snapshot to.
    my $fn = $self->option('debian-snapshot');

    open(my $fh, ">$fn") or $self->fault("can not write to $fn: $!");

    $self->display(1, "Creating snapshot of Debian package installation...");

    my $packages = $self->debian_packages;

    ## write package information to file.
    foreach my $pkg (sort keys %$packages) {
        my $p = $packages->{$pkg};
        print $fh $p->[0], ' ', $p->[1], ' ',
                  $p->[2], ' ', $p->[3], "\n";
    }

    close($fh);

    $self->display(1, "done.\n");

}

sub debian_package_diff ($$$;$)
{
    my $self  = shift->new;

    my $spkgs = shift;
    my $cpkgs = shift;
    my $ipkgs = shift;

    my @install = ();
    my @remove  = ();
    my @purge   = ();

    my $total   = keys(%$spkgs) + keys(%$cpkgs);
    my $done    = 0;
    my $percent = 0;

    ## compare server packages to the client and create
    ## list of apt-get commands to run on the client.
    foreach my $pkg (keys %$spkgs) {

        ## display progress update.
        if ($self->verbose and tty) {
            my $new = int( ($done++ / $total) * 100);
            if ($new > $percent) {
                my $str = "$new%";
                $self->display(1, "$str", "\b" x length($str));
                $percent = $new;
            }
        }

        ## check if we should ignore this package.
        my $ignore = 0;
        foreach my $regex (@$ipkgs) {
            if ($pkg =~ /$regex/) {
                $ignore = 1;
                last;
            }
        }
        next if $ignore;

        my $spkg = $spkgs->{$pkg};

        ## if installed, make sure client installs it.
        if ($spkg->[0] eq 'ii ') {
            ## client has package installed
            if (exists $cpkgs->{$pkg} and $cpkgs->{$pkg}[0] eq 'ii ') {
                ## check version.
                if ($spkg->[2] ne $cpkgs->{$pkg}[2]) {
                    my $ver = $spkg->[2];
                    push @install, "$pkg=$ver";
                }
            }
            ## client does not have package installed.
            else {
                my $ver = $spkg->[2];
                push @install, "$pkg=$ver";
            }
        }

        ## if removed, make sure client removes it.
        elsif ($spkg->[0] =~ /^[rp]c $/) {
            ## client has package installed
            if (exists $cpkgs->{$pkg} and $cpkgs->{$pkg}[0] !~ /^[rp]c $/) {
                push @remove, $pkg;
            }
        }

        ## unexpected state, possibly flagged for removal,
        ## but not yet uninstalled.
        else {
            $self->display(0, "!! server package $pkg $spkg->[2] ",
                              "has unexpected state: $spkg->[0].\n");
        }

        ## mark package as processed.
        $spkg->[5] = $cpkgs->{$pkg}[5] = 1;

    }

    ## now compare remaining client packages to the server.
    foreach my $pkg (keys %$cpkgs) {

        ## display progress update.
        if ($self->verbose and tty) {
            my $new = int( ($done++ / $total) * 100);
            if ($new > $percent) {
                my $str = "$new%";
                $self->display(1, "$str", "\b" x length($str));
                $percent = $new;
            }
        }

        ## skip if we already processed it.
        next if $cpkgs->{$pkg}[5];

        ## check if we should ignore this package.
        my $ignore = 0;
        foreach my $regex (@$ipkgs) {
            if ($pkg =~ /$regex/) {
                $ignore = 1;
                last;
            }
        }
        next if $ignore;

        push @purge, $pkg;
    }

    return \@install, \@remove, \@purge;
}


## apt_cache_versions($pkg) - get available versions of $pkg using apt-cache

sub apt_cache_versions ($)
{
    my $pkg = shift;

    my @apt = ( 'apt-cache', 'showpkg', $pkg );

    open(my $apt_h, '-|') || exec( @apt )
        or die "error '$!' trying to exec: @apt";

    my $versions = 0;

    my %vers = ();

    while (<$apt_h>) {
        chomp;
        ## iterate through output until we get to Versions: line.
        if ($_ =~ /^Versions:/) {
            $versions = 1;
            next;
        }
        next unless $versions;

        ## we are done soon as we hit blank line after Versions:
        last unless length;

        ## get version and where it's available.
        my ($ver, $list) = $_ =~ m!^([^()]+)\(([^()]+)\)!;

        $vers{$ver} = $list;
    }

    close($apt_h);

    return \%vers;
}

sub debian_closest_version
{
    my $self = shift->new;

    my $install = shift;

    my $total   = @$install;
    my $done    = 0;
    my $percent = 0;

    for (my $i = 0; $i < @$install; ++$i) {
        my $inst = $install->[$i];

        ## display progress update.
        if (tty and $self->verbose) {
            my $new = int( ($done++ / $total) * 100);
            if ($new > $percent) {
                my $str = "$new%";
                $self->display(1, "$str", "\b" x length($str));
                $percent = $new;
            }
        }

        my ($pkg, $ver) = split /=/, $inst, 2;

        ## get available versions for this package.
        my $vers = apt_cache_versions($pkg);

        ## use this version if it's available.
        next if $vers->{$ver};

        ## otherwise find the closest available package
        ## that is a newer version.
        my $last_ver = '';
        foreach my $aver (sort { dpkg_cmp_ver($a, $b) } keys %$vers) {
            $last_ver = $aver;
            last if dpkg_cmp_ver($ver, $aver) < 1;
        }

        ## set version if we found one.
        if ($last_ver) {
            $self->display(1, "* using $pkg $last_ver instead of $ver\n");
            $install->[$i] = "$pkg=$last_ver" if $last_ver;
        }
    }
}

sub debian_from_snapshot
{
    my $self = shift->new;

    ## open file to read snapshot from.
    my $fn = $self->option('debian-from-snapshot');
    open(my $fh, "<$fn") or $self->fault("can not read from $fn: $!");
    my $spkgs = parse_package_list($fh);
    close($fh);

    $self->display(1, "Determining current package list...");
    my $cpkgs = $self->debian_packages;

    $self->display(1, "done.\nCalculating differences...");

    my $ipkgs = [ map { split } @{ $self->option('debian-ignore') || [''] } ];

    my ($install, $remove, $purge) =
        $self->debian_package_diff($spkgs, $cpkgs, $ipkgs);

    unless (@$install || @$remove || @$purge) {
        $self->display(1, "No differences.\n");
        return;
    }

    $self->display(1, "out-of-sync.\n");

    ## if debian-closest-version was specified, check each install
    ## version and make sure it is a version that can be installed.
    if (@$install and $self->option('debian-closest-version')) {
        $self->debian_closest_version($install);
    }

    my $dryrun = $self->dryrun ? ' (dry-run) ' : '';

    $self->display(1, "Installing/removing packages with apt-get$dryrun...\n");

    my @apt = 'apt-get';
    push @apt, '--dry-run' if $dryrun;
    push @apt, 'install', @$install;
    push @apt, map { "$_-" } @$remove;
    push @apt, map { "$_-" } @$purge;

    $self->display(0, "dry-run: @apt\n") if $dryrun;

    $self->system_call(@apt)
        or $self->fault("something went wrong running apt-get install: code ", $? >> 8);

    if (@$purge) {
        $self->display(1, "Purging package configurations$dryrun...\n");

        my @dpkg = 'dpkg';
        push @dpkg, '--no-act' if $dryrun;
        push @dpkg, '--purge', @$purge;

        $self->display(0, "dry-run: @dpkg\n") if $dryrun;

        $self->system_call(@dpkg)
            or $self->fault("something went wrong running dpkg --purge: code ", $? >> 8);
    }

    $self->display(1, "Debian package installation syncronized.\n");
}

sub verbose
{
    my $self = shift->new;

    return $self->option('v') || 0;
}

sub timestamp
{
    my $self = shift->new;
    
    return $self->option('t') ? '[%Y-%m-%d %T]' : 0;
}

sub rsync_version {
    my $self  = shift->new;
    my $check = shift;

    unless ($self->{rsync_version}) {
        open(my $fh, "rsync --version|")
            or $self->fault("could not get rsync version: $! (code ", $? >> 8, ")");
        my $ver_line = <$fh>;
        close($fh);
        
        ## sample output.
        ## rsync  version 2.6.4  protocol version 29
        my ($version) = $ver_line =~ /rsync\s+version\s+(\d+\.\d+\.\d+)/;
        $version or $self->fault("could not determine rsync version: $ver_line");
        
        $self->display(3, "  # rsync version: $version\n");
        
        $self->{rsync_version} = $version;
    }
    
    if ($check) {
        no warnings 'uninitialized';
        my @ver = split /\./, $self->{rsync_version};
        my @des = split /\./, $check;
        
        for (my $i = 0; $i < @des; ++$i) {
            return 1 if $ver[$i] > $des[$i];
            return 0 if $ver[$i] < $des[$i];
        }
        return 1;
    }
    
    return wantarray ? (split /\./, $self->{rsync_version})
                     : $self->{rsync_version};
}

sub exit_code {
    my $self = shift->new;
    if (@_) {
        $self->{exit_code} ||= shift;
        return $self;
    }
    return $self->{exit_code} || 0;
}

1;
